README.md
#Provides an ingame technology browser.
It is a help system.
All places where a Game object is referenced are listed with this object.
For example, for an item, all recipes that process it and that create it are displayed. If it is an assembly machine, all recipes that this machine can process are also displayed, and so on.
This is the general principle of this mod.
The displayed icons also show dynamic information. For example, a recipe shows whether it has already been researched and if so, how much handcrafting is possible for the player with the available inventory.
Also, you can trigger certain actions directly from the help system.
- Research
- Handcrafting
- Pipette
There is a reminder panel with optional automatic research and (hand-)crafting.
There is also a history.

It is a tribute to the help system of Evospace. I hope factorio players will like it a lot as well.
The "remidor panel" is inspired by the one from Satisfactory.

###todo: *(please feel free to rate)*
- provide information on recipe-combinations
- involve achievements
- machine-column for labs
- Improve recipe-column-layout in case of huge amount of recipes
- Keep tab-selection for huge presentator after refresh
- Reminder task helpers for complex recipes. For instance "(recursively) add a task for all ingredients"
- Reminder tasks for Recipes
- Reminder tasks for Technologies
- Ping for objects ("Where the hell are there some more iron plates lying around?")
- present multiple items_to_place_this of entity-prototypes
- present limitations for modules
- Reminder panel grouping
- Reminder panel priorities
- Reminder panel load/save
- Show steam-processors
- Show energy consumption/production as tooltip on arrow-sprites in recipe lines
- Reminder: remove researches
- Reminder: indicator for available/not available workers
- Presentator: tooltips for tab-group icons
- Present research-status in tooltip of recipes and technologies
- ... *more suggestions?*


###todo(or probably not): *(please feel free to complain)*
- multiplayer *(See no way do this. Help welcome)*
- copy recipe *(See no way do this. Help welcome)*
- Several filter options
  => by mod *(See no way do this. Help welcome)*

###known issue:
- Presentator: if there are only UsefulLinks, they are not presented
- Multiplayer: Desync issue when loading a save file where reminder panel is open *(help welcome!)*
- When using a research queue handler like the one from sonaxaton, ingteb-research may behave strange. (As a workaround, you can temporarily use the replacement mod [Improved Research Queue with interface](https://mods.factorio.com/mod/sonaxaton-research-queue-with-interface) instead of [Improved Research Queue](https://mods.factorio.com/mod/sonaxaton-research-queue).)
- mod-issue.missing-worker should be checked during matadata-scan already

**Note: This mod is under development. Please be so kind and use the "Discussion" tab to give feedback - especially if you encounter bugs.** Many thanks in advance

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/G2G4BH6WX)


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.5.10
Date: 2023-04-25
  Bugfixes:
    - Fixing crash when creating a new game with Space Exploration mod
---------------------------------------------------------------------------------------------------
Version: 0.5.9
Date: 2023-04-25
  Bugfixes:
    - Fixing crash when using Space Exploration mod
---------------------------------------------------------------------------------------------------
Version: 0.5.8
Date: 2023-04-25
  Changes:
    - [admin] don't append version number to toplevel folder of package (done by hacking justarandomgeek.factoriomod-debug-1.1.22)
---------------------------------------------------------------------------------------------------
Version: 0.5.7
Date: 2022-08-04
  Bugfixes:
    - Fixing crash when making empty-barrel a capsule
  Changes:
    - New warning for strange setup of fixed recipes
---------------------------------------------------------------------------------------------------
Version: 0.5.6
Date: 2022-07-01
  Bugfixes:
    - Fixing crash with 256k-mod
---------------------------------------------------------------------------------------------------
Version: 0.5.5
Date: 2022-06-11
  Bugfixes:
    - Fixing crash with missing functions
---------------------------------------------------------------------------------------------------
Version: 0.5.4
Date: 2022-05-10
  Bugfixes:
    - Fixing Crash on opening (thanks OgrumXD)
---------------------------------------------------------------------------------------------------
Version: 0.5.3
Date: 2022-05-04
  Bugfixes:
    - Fixing duplicate descriptions at tooltips of stack-of-goods
    - Fixing crash when creating a new game
---------------------------------------------------------------------------------------------------
Version: 0.5.2
Date: 2022-05-04
  Bugfixes:
    - Fixing crash with fluid-burning recipe
---------------------------------------------------------------------------------------------------
Version: 0.5.1
Date: 2022-05-04
  Bugfixes:
    - Localization does not work
    - Resources are not recognised correctly
---------------------------------------------------------------------------------------------------
Version: 0.5.0
Date: 2022-05-02
  Changes:
    - refactor for fixing desyncs
    - filter options for presentator
    - user setting: show impossible recipes (otherwise they got filtered)
    - reminder selector: better grouping of workers and recipes
    - [tech] Array class got Length field to be able to handle nil values
    - [tech] Use mod on_research_queue_changed
---------------------------------------------------------------------------------------------------
Version: 0.4.1
Date: 2022-04-11
  Bugfixes:
    - Fixing Remindor data is lost when loading a save file
    - Multiplayer: Mitigating Desync issue when loading a save file where any panel is open
---------------------------------------------------------------------------------------------------
Version: 0.4.0
Date: 2022-04-09
  Bugfixes:
    - Mitigating Desync issue when loading a save file where reminder panel is open
    - Fixing Crash when getting information from mod-defined module-categories
---------------------------------------------------------------------------------------------------
Version: 0.3.33
Date: 2022-04-02
  Bugfixes:
    - fixing multiple lines of the same recipe
    - fixing multiple icons for the same worker
    - mitigating Large recipe columns are jerky during handcrafts or research completion
  Changes:
    - Yellow backgroung for recipes that are currently in research queue
---------------------------------------------------------------------------------------------------
Version: 0.3.32
Date: 2022-03-31
  Bugfixes:
    - fixing crash when opening a fuel item.
---------------------------------------------------------------------------------------------------
Version: 0.3.31
Date: 2022-03-27
  Bugfixes:
    - fixing that fixed recipes are not presented anymore.
  Changes:
    - show mini-recipe panel for improper recipes also
    - show fuel category related information
    - speed information on worker sprites improved
    - show fluid fuel categories and their burners
    - show information concerning modules
---------------------------------------------------------------------------------------------------
Version: 0.3.30
Date: 2022-03-26
  Bugfixes:
    - fixing that hidden recipes for furnances are not presented anymore.
---------------------------------------------------------------------------------------------------
Version: 0.3.29
Date: 2022-03-23
  Changes:
    - Special treatment for zombie recipes
    - Better treatment for hidden recipes. Present them only when used as fixed-recipe somewhere.
---------------------------------------------------------------------------------------------------
Version: 0.3.28
Date: 2022-03-19
  Bugfixes:
    - fixing crash when creating a BoilingRecipe
---------------------------------------------------------------------------------------------------
Version: 0.3.27
Date: 2022-03-18
  Changes:
    - logging of issues found in other mods improved
  Bugfixes:
    - fixing incompatible with se-space-trains
---------------------------------------------------------------------------------------------------
Version: 0.3.26
Date: 2022-03-18
  Changes:
    - pipette improved
  Bugfixes:
    - Remaining debug code removed
    - fixing incompatible Pyanodons Coal Processing
    - fixing incompatible with Industrial Revolution 2
    - fixing incompatible with both aai industry and Krastorio 2
---------------------------------------------------------------------------------------------------
Version: 0.3.25
Date: 2022-03-17
  Changes:
    - Show hidden recipes
    - Missing locales added for domains and categories
  Bugfixes:
    - Missing hidden recipe for water to steam

data.lua
local data_util = require '__flib__.data-util'
local Constants = require "Constants"

local big_size = 64
local small_size = 32
local tiny_size = 24
local frame_action_icons = Constants.GraphicsPath .. "frame-action-icons.png"

data:extend(
    {
    {
        type = "custom-input",
        name = Constants.Key.Main,
        key_sequence = "H",
        include_selected_prototype = true,
    },
    { type = "custom-input", name = Constants.Key.Back, key_sequence = "mouse-button-4" },
    { type = "custom-input", name = Constants.Key.Fore, key_sequence = "mouse-button-5" },

    {
        type = "sprite",
        name = "factorio",
        filename = "__core__/graphics/factorio-icon.png",
        size = 64,
        scale = 0.5,
    },

    {
        type = "sprite",
        name = "hidden-recipe",
        filename = "__core__/graphics/favourite-grey.png",
        size = 64,
        scale = 0.5,
    },


    {
        type = "sprite",
        name = "ingteb",
        filename = "__" .. Constants.ModName .. "__/thumbnail.png",
        size = 64,
        scale = 0.5,
    },
    {
        type = "sprite",
        name = "chemical",
        filename = Constants.GraphicsPath .. "chemical.png",
        size = 64,
        scale = 0.5,
    },
    {
        type = "sprite",
        name = "automatic-recipe",
        filename = Constants.GraphicsPath .. "automatic-recipe.png",
        size = 64,
        scale = 0.5,
    },
    {
        type = "sprite",
        name = "effects",
        filename = Constants.GraphicsPath .. "effects.png",
        size = 32,
        scale = 0.5,
    },
    {
        type = "sprite",
        name = "hide-this-column",
        filename = "__core__/graphics/cancel.png",
        size = 64,
        scale = 0.5,
    },
    {
        type = "sprite",
        name = "items-per-timeunit",
        filename = Constants.GraphicsPath .. "items-per-timeunit.png",
        size = 34,
    },
    {
        type = "sprite",
        name = "setting-is-off",
        filename = "__core__/graphics/no-recipe.png",
        width = 101,
        height = 101,
        scale = 0.6,
    },

    data_util.build_sprite("ingteb_settings_black", { 0, 96 }, frame_action_icons, 32),
    data_util.build_sprite("ingteb_settings_white", { 32, 96 }, frame_action_icons, 32),

    -- Icons copied from game core

    {
        type = "sprite",
        name = "go_to_arrow",
        filename = Constants.GraphicsPath .. "goto-icon.png",
        priority = "medium",
        width = 32,
        height = 32,
        flags = { "icon" },
    },

    {
        type = "sprite",
        name = "close_white",

        filename = Constants.GraphicsPath .. "close-white.png",
        priority = "extra-high-no-scale",
        size = 32,
        scale = 0.5,
        flags = { "gui-icon" },
    },
    {
        type = "sprite",
        name = "close_black",
        filename = Constants.GraphicsPath .. "close-black.png",
        priority = "extra-high-no-scale",
        size = 32,
        scale = 0.5,
        flags = { "gui-icon" },
    },

    {
        type = "sprite",
        name = "search_black",
        filename = Constants.GraphicsPath .. "search-black.png",
        priority = "extra-high-no-scale",
        size = 32,
        scale = 0.5,
        flags = { "gui-icon" },
    },
    {
        type = "sprite",
        name = "search_white",
        filename = Constants.GraphicsPath .. "search-white.png",
        priority = "extra-high-no-scale",
        size = 32,
        scale = 0.5,
        flags = { "gui-icon" },
    },

    {
        type = "sprite",
        name = "slot_icon_robot_material",
        filename = Constants.GraphicsPath .. "slot-robot-material-white.png",
        priority = "medium",
        width = 64,
        height = 64,
        mipmap_count = 3,
        scale = 0.5,
        flags = { "gui-icon" },
    },
    {
        type = "sprite",
        name = "slot_icon_robot_material_black",
        filename = Constants.GraphicsPath .. "slot-robot-material-black.png",
        priority = "medium",
        width = 64,
        height = 64,
        mipmap_count = 3,
        scale = 0.5,
        flags = { "gui-icon" },
    },
    {
        type = "sprite",
        name = "technology_black",
        filename = Constants.GraphicsPath .. "technology-black.png",
        size = 64,
        mipmap_count = 2,
        scale = 0.5,
        flags = { "gui-icon" },
    },
    {
        type = "sprite",
        name = "technology_white",
        filename = Constants.GraphicsPath .. "technology-white.png",
        size = 64,
        mipmap_count = 2,
        scale = 0.5,
        flags = { "gui-icon" },
    },
    {
        type = "sprite",
        name = "trash",
        filename = Constants.GraphicsPath .. "trash.png",
        priority = "extra-high-no-scale",
        size = 32,
        flags = { "gui-icon" },
        mipmap_count = 2,
        scale = 0.5,
    },

    {
        type = "sprite",
        name = "trash_white",
        filename = Constants.GraphicsPath .. "trash-white.png",
        priority = "extra-high-no-scale",
        size = 32,
        scale = 0.5,
        mipmap_count = 2,
        flags = { "gui-icon" },
    },
    {
        type = "sprite",
        name = "color_picker_white",
        filename = Constants.GraphicsPath .. "color-picker-white.png",
        priority = "extra-high-no-scale",
        size = 32,
        scale = 0.5,
        mipmap_count = 2,
        flags = { "gui-icon" },
    },
}
)

data.raw["utility-sprites"].default.factorio = {
    filename = "__core__/graphics/factorio.icon",
    priority = "medium",
    size = 32,
    flags = { "icon" },
}

data.raw["utility-sprites"].default.settings = data_util.build_sprite {
    filename = "__core__/graphics/factorio.icon",
    priority = "medium",
    size = 32,
    flags = { "icon" },
}

data.raw["gui-style"].default["ingteb-flow-centered"] = {
    type = "horizontal_flow_style",
    horizontally_stretchable = "on",
    horizontal_align = "center",
}

data.raw["gui-style"].default["ingteb-recipe-scroll"] = {
    type = "scroll_pane_style", --
    parent = "scroll_pane",
    width = 43 * Constants.MaximumEntriesInRecipeList,
}

data.raw["gui-style"].default["ingteb-main-frame"] = {
    type = "frame_style", --
    scalable = true,
}

data.raw["gui-style"].default["ingteb-flow-right"] = { --
    type = "horizontal_flow_style", --
    horizontally_stretchable = "on",
    horizontal_align = "right",
}

data.raw["gui-style"].default["ingteb-flow-fill"] = { --
    type = "vertical_flow_style", --
    horizontally_stretchable = "on",
}

local default_glow_color = { 225, 177, 106, 255 }
local default_dirt_color = { 15, 7, 3, 100 }
local red_color = { 1, 0, 0, 100 }

local function offset_by_2_rounded_corners_glow(tint_value)
    return {
        position = { 240, 736 },
        corner_size = 16,
        tint = tint_value,
        top_outer_border_shift = 4,
        bottom_outer_border_shift = -4,
        left_outer_border_shift = 4,
        right_outer_border_shift = -4,
        draw_type = "outer",
    }
end

local function sprite17(x, y) return { border = 4, position = { x * 17, y * 17 }, size = 16 } end

data.raw["gui-style"].default["ingteb-light-button"] = {
    type = "button_style",
    parent = "button",
    draw_shadow_under_picture = true,
    size = 40,
    padding = 0,
    default_graphical_set = {
        base = sprite17(0, 1),
        -- shadow = offset_by_2_rounded_corners_glow(default_dirt_color)
    },
    hovered_graphical_set = {
        base = sprite17(2, 1),
        -- shadow = offset_by_2_rounded_corners_glow(default_dirt_color),
        -- glow = offset_by_2_rounded_corners_glow(default_glow_color)
    },
    clicked_graphical_set = {
        base = sprite17(3, 1),
        shadow = offset_by_2_rounded_corners_glow(default_dirt_color),
    },
    selected_graphical_set = {
        base = sprite17(2, 1),
        shadow = offset_by_2_rounded_corners_glow(default_dirt_color),
    },
    selected_hovered_graphical_set = {
        base = sprite17(2, 1),
        shadow = offset_by_2_rounded_corners_glow(default_dirt_color),
        glow = offset_by_2_rounded_corners_glow(default_glow_color),
    },
    selected_clicked_graphical_set = {
        base = sprite17(3, 1),
        shadow = offset_by_2_rounded_corners_glow(default_dirt_color),
    },
    pie_progress_color = { 0.98, 0.66, 0.22, 0.5 },
}

data.raw["gui-style"].default["ingteb-un-button"] = { type = "image_style", size = 40 }
data.raw["gui-style"].default["ingteb-tab"] = { type = "tab_style", size = 40 }

data:extend { { type = "font", name = "ingteb-font18", from = "default", size = 18 } }

data:extend { { type = "font", name = "ingteb-font24", from = "default", size = 24 } }

data:extend { { type = "font", name = "ingteb-font32", from = "default", size = 32 } }

data.raw["gui-style"].default["ingteb-big-tab"] = { type = "tab_style", font = "ingteb-font32" }

data.raw["gui-style"].default["ingteb-big-tab-disabled"] = {
    type = "tab_style",
    font = "ingteb-font32",
    default_graphical_set = { base = { position = { 208, 17 }, corner_size = 8 } },
}

data.raw["gui-style"].default["ingteb-medium-tab"] = {
    type = "tab_style",
    font = "ingteb-font18",
    size = 40,
}

data.raw["gui-style"].default["ingteb-big-label"] = { type = "label_style", font = "ingteb-font24" }
